Update to changelog for release (#487)
CrayLabs/SmartSim: doc/changelog.rst

SmartSim
========

0.6.1
-----

Released on 15 February, 2024

Description

- Fix line-ending related errors on MacOS (ARM64)

> - Update instructions for Apple Silicon

Reviewer: Feel free to drop this from the short description since we are already calling out ARM64 support as a new feature at line 27

- Duplicate DBModel/Script prevention
- Update license to include 2024

> - Patch changes for release prep

Reviewer: I think the only thing is this PR that someone would reading the change log on would be that the TM is now on by default. Can we update this description to
```suggestion
- Telemetry monitor is now active by default
```
or something similar?

> - Manifest: has DB objects refactor

Reviewer: Feel free to drop this PR from the short description, I happen to know this was only touching internal code, so it shouldn't be of much interest to anyone reading the change log

Reviewer: The Patch PR and the Manifest DB Object refactor I mean!! Leave the other two!

- Add support for Mac OSX on Apple Silicon
- Add eval() to remove Torch warnings during testing
- Validate Slurm timing format
- Expose Typehints
- Fix test_logs to prevent generation of dir
- Fix Python Typehint for RunSettings.colocated_db_settings
- Add isort/black check to github actions
- Relax typing extensions required version
- Python 3.11 Support
- Quality of life `smart validate` improvements

> - Update outdated actions

Reviewer: This can probably be dropped from the short description just because it is strictly "our" CI that changed

- Remove Cobalt support
- Enrich logging through context vars
- Upgrade Machine Learning dependencies
- Override sphinx-tabs background color
- Add concurrency group to test workflow
- Fix index when installing torch through smart build

Detailed Notes

- Modify the `git clone` for both Redis and RedisAI to set the line endings to
  unix-style line endings when using MacOS on ARM. (SmartSim-PR482_)
- Separate install instructions are now provided for Mac OSX on x64 vs ARM64 (SmartSim-PR479_)

> - This PR prevents duplicate ML models and scripts names being added to an

Reviewer: Small suggestion
```suggestion
- Prevent duplicate ML models and scripts names being added to an
```

  (SmartSim-PR477_)
- Refactor logic of `Manifest.has_db_objects` to remove excess branching
  and improve readability/maintainability. (SmartSim-PR476_)
- SmartSim can now be built and used on platforms using Apple Silicon
  (ARM64). Currently, only the PyTorch backend is supported. Note that libtorch
  will be downloaded from a CrayLabs github repo. (SmartSim-PR465_)
- Tests that were saving Torch models were emitting warnings.  These warnings
  were addressed by updating the model save test function. (SmartSim-PR472_)
- Validate the timing format when requesting a slurm allocation. (SmartSim-PR471_)
- Add and ship `py.typed` marker to expose inline type hints. Fix
  type errors related to SmartRedis. (SmartSim-PR468_)
- Fix the `test_logs.py::test_context_leak` test that was
  erroneously creating a directory named `some value` in SmartSim's root
  directory. (SmartSim-PR467_)
- Add Python type hinting to colocated settings. (SmartSim-PR462_)
- Add github actions for running black and isort checks. (SmartSim-PR464_)
- Relax the required version of `typing_extensions`. (SmartSim-PR459_)
- Addition of Python 3.11 to SmartSim. (SmartSim-PR461_)

> - Quality of life `smart validate` improvements such as seting `CUDA_VISIBLE_DEVICES`

Reviewer: typo
```suggestion
- Quality of life `smart validate` improvements such as setting `CUDA_VISIBLE_DEVICES`
```

  environment variable within `smart validate` prior to importing any ML deps to
  prevent false negatives on multi-GPU systems. Additionally, moving SmartRedis logs
  from standard out to dedicated log file in the validation temporary directory as well as
  suppress `sklearn` deprecation warning by pinning `KMeans` constructor
  argument. Lastly, move TF test to last as TF may reserve the GPUs it uses.
  (SmartSim-PR458_)
- Some actions in the current GitHub CI/CD workflows were outdated. They were
  replaced with their latest versions. (SmartSim-PR446_)
- As the Cobalt workload manager is not used on any system we are aware of,
  its support in SmartSim was terminated and classes such as `CobaltLauncher` have
  been removed. (SmartSim-PR448_)

> - The implementation makes use of python `contextvars.ContextVar` to store

Reviewer: The implementation of what exactly?

  experiment-specific state. The state is used to dynamically modify
  experiment-level logging. (SmartSim-PR452_)
- Updated SmartSim's machine learning backends to PyTorch 2.0.1, Tensorflow
  2.13.1, ONNX 1.14.1, and ONNX Runtime 1.16.1. As a result of this change,
  there is now an available ONNX wheel for use with Python 3.10, and wheels for
  all of SmartSim's machine learning backends with Python 3.11.
  (SmartSim-PR451_) (SmartSim-PR461_)
- The sphinx-tabs documentation extension uses a white background for the tabs component.
  A custom CSS for those components to inherit the overall theme color has
  been added. (SmartSim-PR453_)
- This PR adds concurrency groups to GitHub's CI/CD workflows, preventing
  multiple workflows from the same PR to be launched concurrently.
  (SmartSim-PR439_)

> - Torch changed something in their indexing when trying to install from
>   their provided wheels. This updates the `pip install` command within

Reviewer: Just to be a bit more specific
```suggestion
- Torch changed their preferred indexing when trying to install their
  provided wheels. Updated the `pip install` command within
```

  `smart build` to ensure that the appropriate packages can be found.
  (SmartSim-PR449_)

.. _SmartSim-PR482: https://github.com/CrayLabs/SmartSim/pull/482
.. _SmartSim-PR479: https://github.com/CrayLabs/SmartSim/pull/479
.. _SmartSim-PR475: https://github.com/CrayLabs/SmartSim/pull/475
.. _SmartSim-PR485: https://github.com/CrayLabs/SmartSim/pull/485

> .. _SmartSim-PR473: https://github.com/CrayLabs/SmartSim/pull/473

Reviewer: Okay, very incredibly pedantic request, but I am insane and need things to be in order, so you're  gonna have to bear with me: can we sort this list of links, hahaha?

Author: yes! I put it in order of merged but can switch
